rs/nns/integration_tests/src/cycles_minting_canister.rs

```diff
@@ -778,6 +784,255 @@ fn test_cmc_cycles_create_with_settings() {
     );
 }
 
+#[test]
+fn test_cmc_automatically_refunds_when_memo_is_garbage() {
+    // Step 1: Prepare the world.
+
+    // USER1 has some ICP (in their default subaccount).
+    let account_identifier = AccountIdentifier::new(*TEST_USER1_PRINCIPAL, None);
+    let initial_balance = Tokens::new(100, 0).unwrap();
+    let state_machine = state_machine_builder_for_nns_tests().build();
+    let nns_init_payloads = NnsInitPayloadsBuilder::new()
+        .with_ledger_account(account_identifier, initial_balance)
+        .build();
+    setup_nns_canisters(&state_machine, nns_init_payloads);
+
+    let assert_canister_statuses_fixed = |narrative_phase| {
+        assert_eq!(
+            btreemap! {
+                btreemap! { "status".to_string() => "running".to_string() } => 11,
+                btreemap! { "status".to_string() => "stopped".to_string() } => 0,
+                btreemap! { "status".to_string() => "stopping".to_string() } => 0,
+            },
+            fetch_int_gauge_vec(
+                state_machine.metrics_registry(),
+                "replicated_state_registered_canisters"
+            ),
+            "{}",
+            narrative_phase
+        );
+    };
+    // This will be called again later to verify that no canisters were added.
+    // Here, we just make sure that assert_canister_statuses_fixed has a correct
+    // understanding of how many canisters there are originally.
+    assert_canister_statuses_fixed("start");
+
+    let assert_balance = |nominal_amount_tokens: u64, fee_count: u64, narrative_phase: &str| {
+        let observed_balance = icrc1_balance(
+            &state_machine,
+            LEDGER_CANISTER_ID,
+            Account {
+                owner: Principal::from(*TEST_USER1_PRINCIPAL),
+                subaccount: None,
+            },
+        );
+
+        let total_fees = Tokens::new(0, fee_count * 10_000).unwrap();
+        let expected_balance = Tokens::new(nominal_amount_tokens, 0)
+            .unwrap()
+            .checked_sub(&total_fees)
+            .unwrap();
+        assert_eq!(observed_balance, expected_balance, "{}", narrative_phase);
+    };
+    // This is more to gain confidence that assert_balance works; there is very
+    // little risk that USER1's balance is not 100.
+    assert_balance(100, 0, "start");
+
+    // Step 2: Run code under test.
+
+    // Step 2.1: Send ICP from USER1 to CMC, but use a garbage memo. There is no
+    // immediate explosion; "the bomb is only being planted" so to speak.
```

Comment:
this language sounds rather alarming, perhaps incendiary :)

It also sounds like a bad memo is much more exciting than it is

Reply:
I asked Chat-Gpt to make a meme about this, here's what it suggested: 
![9hnqpj](https://github.com/user-attachments/assets/30efd1e9-70fe-4850-be8c-9c7a9afc7baa)